File: backends/python/micro_news.py
from flask import Flask, jsonify
from flask_cors import CORS, cross_origin
import requests

app = Flask(__name__)
CORS(app)

API_KEY = '8dade0feadcf43e285e215cc7271de9c'
URL = f'https://newsapi.org/v2/top-headlines?country=us&apiKey={API_KEY}'

def add_cors_headers(response):
    response.headers.add('Access-Control-Allow-Origin', 'http://localhost:3000')
    response.headers.add('Access-Control-Allow-Methods', 'GET')
    response.headers.add('Access-Control-Allow-Headers', 'Content-Type')
    return response

@app.route('/api/getArticle/<title>', methods=['GET'])
def get_article(title):
    response = requests.get(URL)
    data = response.json()
    article = next((a for a in data['articles'] if a['title'] == title), None)
    response = jsonify(article)
    return add_cors_headers(response)

@app.route('/api/getAuthorArticles/<author>', methods=['GET'])
def get_author_articles(author):
    response = requests.get(URL)
    data = response.json()
    articles = [a for a in data['articles'] if a['author'] == author]
    response = jsonify(articles)
    return add_cors_headers(response)

@app.route('/api/getTopHeadlines', methods=['GET'])
def get_top_headlines():        
    response = requests.get(f'https://newsapi.org/v2/top-headlines?country=us&apiKey={API_KEY}')
    response = jsonify(response.json())
    return add_cors_headers(response)

if __name__ == '__main__':
    app.run(port=8080)

File: backends/node/micro_news.js
const express = require("express");
const axios = require("axios");

const app = express();
const port = 8080;

const API_KEY = "8dade0feadcf43e285e215cc7271de9c";
const BASE_URL = "https://newsapi.org/v2";

// Middleware para tratamento de CORS
app.use((req, res, next) => {
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "GET");
  next();
});

// Rota para obter os Top Headlines
app.get("/api/getTopHeadlines", async (req, res) => {
  try {
    const url = `${BASE_URL}/top-headlines?country=us&apiKey=${API_KEY}`;
    const response = await axios.get(url);
    res.json(response.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Failed to fetch top headlines" });
  }
});

// Rota para obter artigos de um autor
app.get("/api/getAuthorArticles/:author", async (req, res) => {
  try {
    const { author } = req.params;
    const url = `${BASE_URL}/everything?language=en&apiKey=${API_KEY}&q=${author}`;
    const response = await axios.get(url);
    res.json(response.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Failed to fetch author articles" });
  }
});

// Rota para obter um artigo pelo título
app.get("/api/getArticle/:title", async (req, res) => {
  try {
    const { title } = req.params;
    const url = `${BASE_URL}/everything?language=en&apiKey=${API_KEY}&qInTitle=${title}`;
    const response = await axios.get(url);
    res.json(response.data);
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: "Failed to fetch article" });
  }
});

// Iniciar o servidor
app.listen(port, () => {
  console.log(`Server running on port ${port}`);
});
